python/cudf/cudf/core/column/struct.py

```diff
@@ -122,14 +125,28 @@ def element_indexing(self, index: int) -> dict:
         result = super().element_indexing(index)
         return self.dtype._recursively_replace_fields(result)
 
-    def __setitem__(self, key, value):
+    def _cast_setitem_value(self, value: Any) -> plc.Scalar:
         if isinstance(value, dict):
-            # filling in fields not in dict
-            for field in self.dtype.fields:
-                value[field] = value.get(field, NA)
 
-            value = cudf.Scalar(value, self.dtype)
-        super().__setitem__(key, value)
+            def _maybe_na_to_none(value: Any) -> Any:
+                if _is_null_host_scalar(value):
+                    return None
+                else:
+                    return value
```

Comment:
We should maybe move this function out of here so that it's not defined inline. It doesn't seem to depend on any state from the outer function, but there is overlap in the function arg which can make things confusing. Also it'll be more expensive to instantiate it every time that this function is called.

Reply:
Sure thing. Move it to the top of the file instead.